feat: transit from jwt-go to go-jose

narg95 opened the pull request:
## Related issue

#514 

## Proposed changes

This pull request migrates from `jwt-go` to `go-jose`. The following are the key considerations for this transition:

- `go-jose` [does not support](https://github.com/square/go-jose/issues/185) `alg:none` therefore tests trying to check this functionality where removed
- `JWTStrategy` interface was kept almost identical to provide an easy migration, except for the following changes:
    * `jwt.Token` struct type from `jwt-go` was replaced by a local and simplified type definition `Token` that covers all `fosite` use cases
    *   `jwt.MapClaims` was replaced by a local copy of that data type
    * `jwt.Claims` interface dependency to `jwt-go` was replaced by the local `MapClaims` struct type and not by another interface because it simplified the adaptation and `fosite` always type casted `jwt.Claims` to `*jwt.MapClaims`. Source code performing this type cast was therefore not needed and removed
- `jwt.ValidationErrors` from `jwt-go` was replaced by a local copy in order to keep backwards compatibility with jwt validation errors
 - `jwt.SigningMethod{Family}` from `jwt-go` has not equivalence to `go-jose` signing methods, therefore it was updated by using concrete signing method names e.g. `jwt.SigningMethodRSA` -> `jose.RS256, jose.RS384, jose.RS512`
- `RS256JWTStrategy` and `ES256JWTStrategy` types were kept to stay backwards compatible, but they are just the same implementation calling under the hood the same helper functions with some minor difference in parameters. This is so because `go-jose` makes no distinction on signing methods, this is hidden from callers and it is inferred based on the private/public key provided.

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md)
      and signed the CLA.
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [x] I have added necessary documentation within the code base (if
      appropriate).

## Further comments

As mentioned in the issue, this is a first step in a full migration to `go-jose`. This step aims to remove `jwt-go` but keeping best-effort-backwards-compatible methods and behaviors using underneaths `go-jose`. This way simplifies the transition process. It is known that `go-jose` offer other benefits, but those need to be addressed in further pull requests.

narg95 commented:
I do not understand why the conformity `OpenID Connect Conformity Tests / oidc-conformity (pull_request) ` is failing. It is also happening in main branch

aeneasr commented:
The problem here is I think one of forks - on master I have to double check. Basically we are doing a module rewrite with the latest PR commit hash in go.mod of Ory Hydra for fosite. This fails apparently for forks. Also GitHub recently changed the way how SHA is computed for PRs as they use a virtual merge commit to do the diff or something. Long story short - probably not related to your changes but we need to fix it before merge. I thought i fixed it but apparently not so for forks

aeneasr commented:
I've created https://github.com/ory/hydra/pull/2525 - it shows how to upgrade fosite to point to your branch. To ensure that OIDC works correctly and there are no serious regressions could you please get Ory Hydra tests to pass? If they're green there we can merge this one here right away! Thank you :)

narg95 commented:
@aeneasr sure, I am on it, so far it compiles, linter is happy, I am taking a look to the tests. I had to upgrade my go to 1.6 to use `import "embed"`. `make docker` works. So it looks promising :)

aeneasr commented on authorize_request_handler_oidc_request_test.go:
```diff
@@ -190,18 +182,6 @@ func TestAuthorizeRequestParametersFromOpenIDConnectRequest(t *testing.T) {
 			expectForm: url.Values{"response_type": {"token"}, "response_mode": {"post_form"}, "scope": {"foo openid"}, "request_uri": {reqTS.URL}, "foo": {"bar"}, "baz": {"baz"}},
 		},
-		{
-			d:          "should pass when request object uses algorithm none",
```
A particular reason for removing these? These are neccessary for some OIDC flows :/

narg95 replied:
I removed them because go-jose [does not support](https://github.com/square/go-jose/issues/185) alg:none therefore tests trying to check this functionality where removed. I assumed it was just a sanity check, alg:none is seen as insecure and a door for vulnerabilities. But if required I can provide that functionality. It is just to base64-urlencoded `header.claims.empty-signature`, right?

aeneasr replied:
Yeah, unfortunately it's still supported as an algorithm in OIDC for whatever reason (probably some legacy corp not being able to implement JWT with algs) - let's see the result of https://github.com/ory/hydra/pull/2526/checks?check_run_id=2554759989 and if it passes we can leave this removed

aeneasr reviewed (commented):
I noticed that `none` was removed from the test assertions - it's unfortunately needed to pass OIDC certification iirc.

I started a test run for OIDC certification here: https://github.com/ory/hydra/pull/2526/checks?check_run_id=2554759989

Let's see if it passes :)

aeneasr commented:
As suspected, the OIDC tests fail with the none alg missing: https://github.com/ory/hydra/pull/2526/checks?check_run_id=2554759989#step:5:114

aeneasr commented:
Thank you! I approved the run just now. It's really annoying that GitHub requires us to hit "Approve Run" over and over again :( Especially if the task takes an hour or more to finish...

aeneasr commented on token/jwt/token.go:
```diff
@@ -0,0 +1,229 @@
+// Parse, validate, and return a token.
+// keyFunc will receive the parsed token and should return the key for validating.
+// If everything is kosher, err will be nil
+func ParseWithClaims(rawToken string, claims MapClaims, keyFunc Keyfunc) (*Token, error) {
```
I think we need tests for this, as we're using it to verify authentication tokens and it looks like it has several moving parts. I think as part of the tests we should have:

1. Completely invalid token like`asdf`
2. Token that is not a JWT `foo.bar.baz`
3. Token that mismatches the keyfunc:
   1. Expect RSA256 but get RSA512
   2. Expect RSA256 but get None
5.  RSA256 Token with unknown signing key
6. Token with invalid claims:
   1. Expired
   1. Incorrect IAT
   2. ...and other cases that should be covered by this

And finally also some passing tokens like:

1. RSA
2. None
3. ECDSA

narg95 replied:
done

aeneasr reviewed (commented):
Awesome, thank you for the great work! It looks like the CI in Ory Hydra is passing so this is a ✅ 😃 

I think we need a couple more tests in here but otherwise this looks really solid! I'll go over the jwt token validation strategy once again when tests have been added.

All in all thank you for the great contribution!

aeneasr commented:
I've updated the CI config so it now runs the conformity tests directly here!

narg95 commented on authorize_request_handler_oidc_request_test.go:
```diff
@@ -217,10 +217,15 @@ func TestAuthorizeRequestParametersFromOpenIDConnectRequest(t *testing.T) {
 				}
 			} else {
-				require.NoError(t, err)
+				if err != nil {
+					real := new(RFC6749Error)
+					errors.As(err, &real)
+					require.NoErrorf(t, err, "Hint: %v\nDebug:%v", real.HintField, real.DebugField)
+				}
```
kind of, it helps to print more details about an unexpected error to find its cause easier :)
It helped me

narg95 commented on token/jwt/token_test.go:
```diff
@@ -0,0 +1,47 @@
+package jwt
```
yes, I took most of `parse_test.go` test cases, I prune those we do not need or support e.g. `json.Number` and added specific code to prevent regresses like returning always a toke instance and returning a specific validation error when a Key was not provided, the rest worked perfect. I also appended the test cases that you mentioned.

narg95 commented on token/jwt/map_claims.go:
```diff
@@ -0,0 +1,154 @@
+package jwt
```
This one did not have explicit tests, it was tested implicitly in `jwt-go` in `parse_test.go`. Those tests I copied.

narg95 commented:
The OpenID conformity tests are broken because they run based on hydra `main` branch which is incompatible.
I have updated `hydra` `go-jose_migration` branch to point to latest commit here, so we can double check that conformity tests pass.

narg95 commented:
I hope those tests provide the confidence required to merge it 🤞 
This PR is becoming bigger ;)

narg95 commented on .github/workflows/oidc-conformity.yml:
```diff
@@ -13,8 +13,8 @@ jobs:
           fetch-depth: 2
-          repository: ory/hydra
-          ref: master
+          repository: narg95/hydra
```
This is a small hack :) to get the OpenID conformity test running in sync with the migration branch used in hydra PR.
This needs to be reverted before merging

narg95 commented:
hi @aeneasr, perhaps you did not noticed it :) but I already addressed your comments and conformity tests are green, so from my point of view this is ready to be merged. Could you take a look?

Btw. before or after merging the `.github/workflows/oidc-conformity.yml` file must be reverted to point to `ory/hydra` repo and ` master` branch, as I explained above, this was needed to run conformity tests with the correct hydra's go-migration branch.

aeneasr commented on token/jwt/map_claims.go:
```diff
@@ -0,0 +1,154 @@
+	if !ok {
+		strAud, ok := m["aud"].(string)
+		if !ok {
+			return false
```
If I read this correctly, this would fail when `aud` is not set and `req` is `false` which is contrary to what `If required is false, this method will return true if the value matches or is unset` says.

I think a test for this would be good!

narg95 replied:
Good catch, agree, actually the error is in the forked `https://github.com/form3tech-oss/jwt-go/blob/v3.2.1/map_claims.go` and not in `https://github.com/dgrijalva/jwt-go/blob/v3.2.0/map_claims.go`. I will double check and update the PR with the fix and some tests with a new commit so you do not have to review all again.

narg95 replied:
I updated with latest `map_claims.go` from `form3tech-oss/jwt-go` which accepts `aud` as an array. I noted that repo provides `map_claims_test.go`, I copy them and added a test to check for the inconsistency that you found.

aeneasr commented on token/jwt/token_test.go:
```diff
@@ -0,0 +1,379 @@
+			errors:      ValidationErrorMalformed,
+		},
+		{
+			name:        "wrong key, expected RSA got ECDSA",
```
This appears to be

```suggestion
			name:        "wrong key, expected ECDSA got RSA",
```

narg95 replied:
Good catch, fixed, I structured a bit better the tests to make it more explicit what is given and what is expected, I copied tests from `dgrijalva/jwt-go` were a bit confusing.

aeneasr commented on token/jwt/token_test.go:
```diff
@@ -0,0 +1,379 @@
+			errors:      ValidationErrorSignatureInvalid,
+		},
+		{
+			name:        "nokey, expected RSA got ECDSA",
```
This appears to be

```suggestion
			name:        "nokey, got RSA but found no key",
```

narg95 replied:
fixed

aeneasr commented on token/jwt/token_test.go:
```diff
@@ -0,0 +1,379 @@
+			errors:      ValidationErrorSignatureInvalid,
+			signingKey:  UnsafeAllowNoneSignatureType,
+			method:      SigningMethodNone,
+		},
```
Could you please add one more test here:

1. `should fail, expected ECDSA but got NONE` which is the invert to https://github.com/ory/fosite/pull/593/files#diff-62425fcddec1b39f6b477e442647c937f444850b79d1acdbfc8b2ad7289d7a31R210

narg95 replied:
done

aeneasr reviewed (commented):
Thank you! This looks much better already :) Sorry for the slow process in this PR. My workload is so intense at the moment and these type of PRs (that touch cryptography) are always very difficult to review!

I'm leaving for holiday on saturday for ~10 days but I think we can get it merged before :)

narg95 commented:
> Thank you! This looks much better already :) Sorry for the slow process in this PR. My workload is so intense at the moment and these type of PRs (that touch cryptography) are always very difficult to review!
> 
> I'm leaving for holiday on saturday for ~10 days but I think we can get it merged before :)

It is understandable, it touches a core component. I think we can manage a merge before Saturday, I think I can get the changes by tomorrow

aarmam commented:
> Thank you! This looks much better already :) Sorry for the slow process in this PR. My workload is so intense at the moment and these type of PRs (that touch cryptography) are always very difficult to review!
> 
> I'm leaving for holiday on saturday for ~10 days but I think we can get it merged before :)

That would be great! I'm looking to add HSM support and this PR would be greatly appreciated!

aeneasr commented on .github/workflows/oidc-conformity.yml:
```diff
@@ -13,8 +13,8 @@ jobs:
           fetch-depth: 2
-          repository: ory/hydra
-          ref: master
+          repository: narg95/hydra
+          ref: go-jose_migration
```
```suggestion
          repository: ory/hydra
          ref: master
```

aeneasr reviewed (approved):
Awesome, good job! Will release this as the next minor version :)

mitar commented:
It is worth noting that now JWT tokens do not unclude `typ` anymore in the header: https://github.com/square/go-jose/issues/126

aeneasr commented:
https://github.com/ory/fosite/issues/606

mitar commented on token/jwt/map_claims.go:
```diff
@@ -0,0 +1,162 @@
+
+	if !m.VerifyExpiresAt(now, false) {
+		vErr.Inner = errors.New("Token is expired")
+		vErr.Errors |= ValidationErrorExpired
```
@ narg95 One late question about this change. So I was looking into this codebase while debugging on our side and I noticed that `Errors` is a bitset value, where multiple errors can be set. I like that. But, everywhere else code is using `==` to test for various types of errors like in the switch/case case, where it is `case jwt.ValidationErrorExpired` for example, that switch/case would reach the end if there are two errors set at any point, no?

The `ve.Errors == jwt.ValidationErrorExpired` check which is there is in fact OK, because that case should be handled only if only `ValidationErrorExpired` is set.
